### ISSUE TITLE
Generated HTML files do not have stable names across recent versions

### ISSUE THREAD

**Issue by wismill** [doxygen/doxygen#10146]
**Describe the bug**
The generated HTML files do not have stable names across versions.

**Expected behavior**
The generated files should have names independent of the Doxygen version, or the changes should be rare and documented in release notes.

**To Reproduce**
Use the [archive](https://github.com/xkbcommon/libxkbcommon/archive/f3210cbf27c79ed4294f85271476c6433446102f.zip) of the project [xkbcommon](https://github.com/xkbcommon).

```bash
meson setup build  -Denable-wayland=false -Denable-x11=false -Denable-docs=true -Denable-tools=false
# Observe the result of the following command in build/html with different versions of Doxygen.
# For example, the output corresponding to `doc/introduction-to-xkb.md` is named:
# - xkb-intro.html in 1.9.6
# - md_doc_2introduction-to-xkb.html in 1.9.7
# - xkb-intro.html in master
# For example, the output corresponding to `doc/quick-guide.md` is named:
# - md_doc_quick_guide.html in 1.9.6
# - md_doc_2quick-guide.html in 1.9.7
# - md_doc_2quick-guide.html in master
meson compile -C doc
```

**Version**
- 1.9.6 (system)
- 19.7 (compiled myself from github repo)
- master (compiled myself from github repo)

Platform: Linux 64 bit, openSUSE Tumbleweed

**Additional context**
In the [xkbcommon](https://github.com/xkbcommon) project we try to follow [“Cool URIs don’t change”](https://www.w3.org/Provider/Style/URI.html) guidelines. This allow us to refactor the documentation without breaking links. In order to do this, we use a [script](https://github.com/xkbcommon/libxkbcommon/blob/master/scripts/ensure-stable-doc-urls.py) that detect file names changes and create explicit redirection pages. This is only semi-automated though, as we rely on the stability of the Doxygen generated filenames.

See the corresponding [issue in xkbcommon](https://github.com/xkbcommon/libxkbcommon/issues/347).

**Comment by albert-github** [comment]
Doxygen tries not to change the URLs either but sometimes it is necessary due to extra functionality, bugs in old implementations etc. The situation you mention regarding the markdown files is unfortunately one of these situations in this respect, there have been some problems in respect to automatic name generation (see the setting `MARKDOWN_ID_STYLE`) and the use of the Header Id attributes(see manual in respect to `## Header 2 ## {#labelid2}`).

It is true that these changes probably should have been mentioned in the release notes but due to the limited time and resources of volunteers it is sometimes hard and also it is never advised (for any project) to use internal generated links etc. To have stable links one could look into the `\anchor`command, but the problem with changed file names remains.

I think we don*t have a problem here but something unfortunate and should be kept in mind for future changes.

**Comment by wismill** [comment]
I understand changes are sometimes required and welcomed if they fix issues. But it worries me that it affects _2 versions in a row_. I can fix it for one version, but I will not do it for each version of Doxygen.

Could you fix the 1.9.7 branch or master to ensure the file names remain the same between 1.9.7 and master?

I think there should be a regression test for this kind of issue. I could contribute to one if you guide me.

**Comment by wismill** [comment]
> it is never advised (for any project) to use internal generated links etc. To have stable links one could look into the \anchorcommand, but the problem with changed file names remains.

I do use the Doxygen mechanisms. The issue is _not_ that the links _inside_ documentation are broken; the issue is that if I share a link in a _message_ (social network, email, Github, etc.) I want this link to remain valid: see [“Cool URIs don’t change”](https://www.w3.org/Provider/Style/URI.html) guidelines. Our script maintains redirection pages (e.g. if we changed the title of a page), but it requires that filenames are stable across Doxygen version.
